[persist/expr] Fallback and optimize for worst-case interpreter path

bkirwi opened the pull request:
The `ColumnSpecs` abstract interpreter has a exponential-time worst-case for function evaluation. In brief: each argument has perhaps four "special" values -- null, errors, range bounds -- and the interpreter needs to evaluate the function with every combination of those values for each argument, which makes the overall runtime exponential in the length of the argument list. Normally this is fine: for binary functions, for example, the exponent is 2, and this is still plenty fast. But this can get expensive for variadic functions with medium-to-long argument lists.

This PR addresses the issue in two ways:
- Adding a new linear-time implementation for associative functions. (If a function is associative, we effectively re-associate it to be evaluated as a linear number of binary function calls, each of which is sufficiently fast.) Conveniently, functions already have an `is_associative` method, and all the functions we care about are included. 
- Limiting the number of arguments for the non-optimized case. This prevents both deep stacks (as seen in https://github.com/MaterializeInc/materialize/issues/19399) and lots of calls to the function itself.

So the common case is now fast, and the uncommon case makes the approximation worse but does not blow up.

### Motivation

Fixes a recognized bug: https://github.com/MaterializeInc/materialize/issues/19400

(I believe it will also fix https://github.com/MaterializeInc/materialize/issues/19399 and https://github.com/MaterializeInc/materialize/issues/19402, but I need to confirm it.)

### Tips for reviewer

Hiding whitespace on this diff makes it a bit more clear!

I was split on whether to log when the newly-introduced limit is hit. It seems like it would be pretty noisy, and `explain with (mfp_pushdown)` will hopefully make it clear that pushdown is not kicking in for their large function calls if they ever expect it, which seems unlikely. But I am open to persuasion here!

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

danhhz reviewed (commented):
approach makes sense to me, but I'll leave the stamp to the experts

aalexandrov commented:
@philip-stoev can you re-run your Tuesday tests and report the before-and-after and report execution time with and without filter pushdown enabled?

philip-stoev reviewed (approved):
This fixes : #19400 and #19399 . However, I think #19402 deserves a separate look -- that bug was about why, if one dataflow is doing busy work, the other dataflows are blocked and the other CPU cores are not used.

This PR was delivered without tests of any kind. I pushed a commit to this PR that removes the dumbing-down of the "limits" test that was necessitated by this bug, so this can be said to constitute a test of sorts for this bug.

However, this PR is missing a test where the fallback strategy is kicking in. No matter how much I pushed it, the explain plan always indicated that pushdown is in effect.

bkirwi commented:
> I pushed a commit to this PR that removes the dumbing-down of the "limits" test that was necessitated by this bug, so this can be said to constitute a test of sorts for this bug.

Thanks!

> However, this PR is missing a test where the fallback strategy is kicking in. No matter how much I pushed it, the explain plan always indicated that pushdown is in effect.

Right. Unfortunately for this test (but fortunately for users) there are currently no variadic functions that are monotonic / "pushdownable" but not associative, so this PR never changes the explain output in practice. I've added a test that confirms that long argument lists are still pushdownable for an associative function, though.

We can test the behaviour of the interpreter at a lower level via the datadriven tests, and confirm that the interpreter "gives up" when it is passed too many arguments. I've added some tests for that as well.

bkirwi commented:
If anyone else is planning to have a look, I'm hoping to get this one in before tomorrow's release. (Thankfully a short one!)

ggevay commented:
I'm taking a look now.

ggevay commented on src/expr/src/interpret.rs:
```diff
@@ -734,6 +740,14 @@ pub struct ColumnSpecs<'a> {
+    // - Adding a linear-time optimization for associative functions like AND, OR, COALESCE.
+    // - Limiting the number of arguments we'll pass on to eval. If this limit is crossed, we'll
+    //   return our default / safe overapproximation instead.
+    const MAX_EVAL_ARGS: usize = 8;
```
I'm wondering how the constant value was determined. If I understand correctly, with this value there might be 4^7 = 2^14 = 16384 evaluations at most. This seems probably ok, but I don't really have an accurate intuition about the cost of a single evaluation.

A related question: nowadays, the optimizer runs on a separate thread from the  environmentd main thread. Does the MFP pushdown also run on the optimizer thread? If on the environmentd main thread, then we have to be extra careful here.

(See a similar magic constant [here](https://github.com/MaterializeInc/materialize/blob/6625ba124f12734408c57715338cabf1802b833a/src/transform/src/literal_constraints.rs#L563), where I aimed for at most 10ms per query.)

philip-stoev replied:
`top` indicates that this work happens in `clusterd` , if I recall correctly. All other queries running on the same compute cluster are blocked for the duration.

bkirwi replied:
That's right - this happens effectively just before the persist client would otherwise fetch a part, to decide whether it's safe to skip it.

bkirwi commented:
Sure - I'll lower it to 6. (The optimized path means that real-world behaviour shouldn't depend too much on the value here, so we might as well be conservative.)

mgree commented on test/sqllogictest/filter-pushdown.slt:
```diff
@@ -140,3 +140,32 @@ Source materialize.public.events
+# Verify explain behaviour for functions with many arguments. In theory, we can't push down
+# non-associative functions with long argument lists... but in practice all the functions we
+# can push down are also associative, so this is moot. Let's at least check that an associative
+# function _does_ report pushdown even when the argument list is long.
```
nit: Wasn't immediately obvious to me that the variadic thing here is the `AND`.

mgree replied:
Just saw you merged. Definitely not worth changing anything here!

bkirwi replied:
I'll roll it up with some other documentation changes!

mgree reviewed (commented):
This looks good to me!
